packages/foam-core/src/model/uri.ts

```diff
@@ -0,0 +1,508 @@
+// Some code coming from
+// https://github.com/microsoft/vscode/tree/master/src/vs/base/common/uri.ts
+
+import * as paths from 'path';
+import GithubSlugger from 'github-slugger';
+import { hash } from '../utils/core';
+import { statSync } from 'fs';
+import { CharCode } from '../common/charCode';
+import { isWindows } from '../common/platform';
+
+/**
+ * Uniform Resource Identifier (URI) http://tools.ietf.org/html/rfc3986.
+ * This class is a simple parser which creates the basic component parts
+ * (http://tools.ietf.org/html/rfc3986#section-3) with minimal validation
+ * and encoding.
+ *
+ * ```txt
+ *       foo://example.com:8042/over/there?name=ferret#nose
+ *       \_/   \______________/\_________/ \_________/ \__/
+ *        |           |            |            |        |
+ *     scheme     authority       path        query   fragment
+ *        |   _____________________|__
+ *       / \ /                        \
+ *       urn:example:animal:ferret:nose
+ * ```
+ */
+export interface URI {
+  scheme: string;
+  authority: string;
+  path: string;
+  query: string;
+  fragment: string;
+}
+
+const { posix } = paths;
+const _empty = '';
+const _slash = '/';
+const _regexp = /^(([^:/?#]+?):)?(\/\/([^/?#]*))?([^?#]*)(\?([^#]*))?(#(.*))?/;
+
+export const create = (
+  scheme: string,
+  authority: string,
+  path: string,
+  query: string,
+  fragment: string
+): URI => ({
+  scheme,
+  authority,
+  path,
+  query,
+  fragment,
+});
+
+export const parse = (value: string): URI => {
+  const match = _regexp.exec(value);
+  if (!match) {
+    return create(_empty, _empty, _empty, _empty, _empty);
+  }
+  return create(
+    match[2] || 'file',
+    percentDecode(match[4] || _empty),
+    percentDecode(match[5] || _empty),
+    percentDecode(match[7] || _empty),
+    percentDecode(match[9] || _empty)
+  );
+};
+
+/**
+ * Parses a URI from value, taking into consideration possible relative paths.
+ *
+ * @param reference the URI to use as reference in case value is a relative path
+ * @param value the value to parse for a URI
+ * @returns the URI from the given value. In case of a relative path, the URI will take into account
+ * the reference from which it is computed
+ */
+export const parseWithReference = (value: string, reference: URI): URI => {
```

Comment:
Would consider naming this `resolve` as in the node path.resolve API: https://nodejs.org/api/path.html#path_path_resolve_paths